### PR TITLE
Add support for class customization via config

### LINKED ISSUE #116: Customize generated class
Hi and thank you for this plugin.

To support our use case, we need the generated class to have an extra annotation.
Could you please add a way to customize the output?

I can see in `CompanionGenerator.java` that using a templating engine is deemed overkill.

For us, something as simple as specifying a base class would do it, but for others new issues may arise, eg: what if specific features require to inherit from specific parent class...

Thanks!

### PR DESCRIPTION
**Pull request by leonard84**
fixes #116

### CODE REVIEW

File: maven-plugin/src/main/java/com/gradle/cucumber/companion/maven/GenerateCucumberCompanionMojo.java [hunk 1]
```diff
@@ -57,9 +60,27 @@ public class GenerateCucumberCompanionMojo extends AbstractMojo {
...
         generatedSourcesDirectoryInternal = Paths.get(generatedSourcesDirectory);
+        generatedClassOptions = customizeGeneratedClasses == null
+            ? new GeneratedClassOptions(
+            Optional.empty(),
+            Collections.emptyList(),
+            Collections.emptyList(),
+            allowEmptySuites
+        )
+            : new GeneratedClassOptions(
+            Optional.ofNullable(customizeGeneratedClasses.getBaseClass()).filter(it -> !it.trim().isEmpty()),
```
Comment (pshevche) [inline maven-plugin/src/main/java/com/gradle/cucumber/companion/maven/GenerateCucumberCompanionMojo.java:79]: ❔ Do we need to sanitize interfaces and annotations the same way?

**Review by pshevche — APPROVED** [review]
Great stuff!

**Review by jean-andre-gauthier — APPROVED** [review]

**Review by skuzzle — APPROVED** [review]

File: companion-generator/src/main/java/com/gradle/cucumber/companion/generator/GeneratedClassOptions.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        this.baseClass = baseClass;
+        this.interfaces = copyList(interfaces);
+        this.annotations = copyList(annotations);
+        this.allowEmptySuites = allowEmptySuites;
+    }
+
+    private static List<String> copyList(List<String> list) {
+        return list.isEmpty()
+            ? Collections.emptyList()
+            : Collections.unmodifiableList(new ArrayList<>(list));
```
[the remark this answers is not in the document]
Comment (leonard84) [reply companion-generator/src/main/java/com/gradle/cucumber/companion/generator/GeneratedClassOptions.java:40]: Yes, it's a general defensive programming technique. We don't control the input, and it is potentially mutable. We could optimize it by checking for Kotlin's and other's immutable types, but I don't think it is warranted.

File: gradle-plugin/src/main/kotlin/com/gradle/cucumber/companion/CucumberCompanionHelper.kt [hunk 3]
```diff
@@ -65,6 +68,10 @@ fun generateCucumberSuiteCompanion(
         // this is a bit icky, ideally we'd use a SourceDirectorySet ourselves, but I'm not sure that is proper
         this.cucumberFeatureSources.set(sourceSet.resources.srcDirs.first())
         this.outputDirectory.set(outputDir)
+        allowEmptySuites.convention(extension.allowEmptySuites)
+        customizeGeneratedClasses.baseClass.convention(extension.customizeGeneratedClasses.baseClass)
+        customizeGeneratedClasses.annotations.convention(extension.customizeGeneratedClasses.annotations)
+        customizeGeneratedClasses.interfaces.convention(extension.customizeGeneratedClasses.interfaces)
```
[the remark this answers is not in the document]
Comment (leonard84) [reply gradle-plugin/src/main/kotlin/com/gradle/cucumber/companion/CucumberCompanionHelper.kt:74]: AFAIK, `convention` should be used for the `default` value. I'd consider the extension's values to be the defaults in this case.

File: gradle-plugin/src/main/kotlin/com/gradle/cucumber/companion/CucumberCompanionExtension.kt [hunk 4]
```diff
@@ -41,28 +45,28 @@ abstract class CucumberCompanionExtension @Inject constructor(
...
     abstract val allowEmptySuites: Property<Boolean>
 
-    /**
-     * Configuration falls back to values, configured at the extension level.
-     */
-    private val defaultConfigureTask = Action<GenerateCucumberSuiteCompanionTask> {
-        allowEmptySuites.set(this@CucumberCompanionExtension.allowEmptySuites.get())
-    }
+    @get:Nested
+    val customizeGeneratedClasses = objects.newInstance(GeneratedClassCustomization::class.java)
```
[the remark this answers is not in the document]
Comment (leonard84) [reply gradle-plugin/src/main/kotlin/com/gradle/cucumber/companion/CucumberCompanionExtension.kt]: right, fixed

**Pull request merged**